### DISCUSSION TITLE
Difference between `yield_per` and `partitions`

### DISCUSSION THREAD

**Discussion by chrisjsewell** [sqlalchemy/sqlalchemy#7030] [Usage Questions]
Heya, I wasn't quite clear from https://docs.sqlalchemy.org/en/14/orm/queryguide.html?highlight=yield%20per#yield-per,
what is the difference between:

```python
>>> stmt = select(User).execution_options(yield_per=10)
>>> for row in session.execute(stmt):
...    print(row)
```

```python
>>> stmt = select(User).execution_options(yield_per=10)
>>> for partition in session.execute(stmt).partitions(10):
...     for row in partition:
...         print(row)
```

How do they affect interactions with the database and performance / memory usage?
Are there use cases to use only yield_per, or is it useless without partitions?

Also, what one is equivalent to (this is what was currently in the codebase I'm updating):

```python
>>> for row in query.yield_per(10):
...     print(row)
```

**Comment by FeehGb** [comment]
I've been using yield_per recently and can confirm that the most efficient way was to use it like this:

```python
for row in query.yield_per(1000).partitions():
    print(row)
```

I saw improvements in both memory usage and performance.

**Comment by CaselIT** [comment]
Hi,

yield_per controls the fetch many size and uses server side cursor where possible, while partitions influences how the iteration happens, the idea being to avoid fetch-one that may not be as optimized as fetch many.

Documentation on version 2 may explain things a bit better: https://docs.sqlalchemy.org/en/20/orm/queryguide/api.html#orm-queryguide-yield-per
